File: zoomimage/src/main/java/mcgars/com/zoomimage/fabric/AnimatorBuilder.java
package mcgars.com.zoomimage.fabric;

import android.support.v4.view.ViewPager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;

import com.alexvasilkov.gestures.transition.SimpleViewsTracker;
import com.alexvasilkov.gestures.transition.ViewsTransitionBuilder;

import java.util.List;

import mcgars.com.zoomimage.ui.ZoomHolder;

/**
 * Created by gars on 18.10.2016.
 */

public class AnimatorBuilder implements AMBuilder {

    private ViewsTransitionBuilder<Integer> builder;

    public static AnimatorBuilder getInstance() {
        return new AnimatorBuilder();
    }

    public ViewsTransitionBuilder<Integer> getBuilder() {
        return builder;
    }

    public AMBuilder from(final ImageView imageView) {
        setBuilder(new ViewTransitionBuilder<Integer>()
                .fromImageView(imageView, new SimpleViewsTracker() {
                    @Override
                    public View getViewForPosition(int position) {
                        return imageView;
                    }
                }));
        return this;
    }

    public AMBuilder from(final List<ImageView> imageView) {
        setBuilder(new ViewTransitionBuilder<Integer>()
                .fromImageView(imageView, new SimpleViewsTracker() {
                    @Override
                    public View getViewForPosition(int position) {
                        return imageView.get(position);
                    }
                }));
        return this;
    }

    public AMBuilder from(final ViewPager fromViewPager) {
        setBuilder(new ViewTransitionBuilder<Integer>()
                .fromViewPager(fromViewPager, new SimpleViewsTracker() {
                    @Override
                    public View getViewForPosition(int position) {
                        return ((ZoomHolder)(fromViewPager.getAdapter())).getImage(position);
                    }
                }));
        return this;
    }

    public AMBuilder from(final RecyclerView recyclerView) {
        setBuilder(new ViewTransitionBuilder<Integer>()
                .fromRecyclerView(recyclerView, new SimpleViewsTracker() {
                    @Override
                    public View getViewForPosition(int position) {
                        ZoomHolder holder = (ZoomHolder) recyclerView.findViewHolderForAdapterPosition(position);
                        return holder == null ? null : holder.getImage(position);
                    }
                }));
        return this;
    }

    protected void setBuilder(ViewsTransitionBuilder<Integer> builder) {
        this.builder = builder;
    }
}

File: zoomimage/src/main/java/mcgars/com/zoomimage/fabric/AnimatorBuilder.kt
package mcgars.com.zoomimage.fabric

import android.view.View
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager.widget.ViewPager
import com.alexvasilkov.gestures.transition.SimpleViewsTracker
import com.alexvasilkov.gestures.transition.ViewsTransitionBuilder
import mcgars.com.zoomimage.ui.ZoomHolder

class AnimatorBuilder private constructor(
        private val builder: ViewsTransitionBuilder<Int>
) : AMBuilder {

    companion object {

        @JvmStatic
        fun from(imageView: ImageView): AMBuilder {
            return AnimatorBuilder(ViewTransitionBuilder<Int>()
                    .fromImageView(imageView, object : SimpleViewsTracker() {
                        override fun getViewForPosition(position: Int): View {
                            return imageView
                        }
                    }))
        }

        @JvmStatic
        fun from(imageView: List<ImageView>): AMBuilder {
            return AnimatorBuilder(ViewTransitionBuilder<Int>()
                    .fromImageView(imageView, object : SimpleViewsTracker() {
                        override fun getViewForPosition(position: Int): View {
                            return imageView[position]
                        }
                    }))
        }

        @JvmStatic
        fun from(fromViewPager: ViewPager): AMBuilder {
            return AnimatorBuilder(ViewTransitionBuilder<Int>()
                    .fromViewPager(fromViewPager, object : SimpleViewsTracker() {
                        override fun getViewForPosition(position: Int): View {
                            return (fromViewPager.adapter as ZoomHolder).getImage(position)
                        }
                    }))
        }

        @JvmStatic
        fun from(recyclerView: RecyclerView): AMBuilder {
            return AnimatorBuilder(ViewTransitionBuilder<Int>()
                    .fromRecyclerView(recyclerView, object : SimpleViewsTracker() {
                        override fun getViewForPosition(position: Int): View? {
                            val holder = recyclerView.findViewHolderForAdapterPosition(position) as ZoomHolder?
                            return holder?.getImage(position)
                        }
                    }))
        }
    }

    override fun build(): ViewsTransitionBuilder<Int> {
        return builder
    }

}
